### PR TITLE
feature/schedules

### PR DESCRIPTION
**Pull request by tegan-temporal**
## Description & motivation 💭 

This is the feature branch for the schedules ui update. 5 separate PRs were merged into this one.

### CODE REVIEW

File: src/lib/components/schedule/schedule-form/spec-type-cron.svelte [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      (v) => ($form.specs[index] = { ...$form.specs[index], cronString: v })
+    }
+    placeholder="* * * * *"
+    required
+    error={!!$errors.specs?.[index]?.cronString?.[0]}
+    hintText={$errors.specs?.[index]?.cronString?.[0] ??
+      translate('schedules.cron-format-hint')}
+  />
+  <button
+    class="mr-auto underline hover:text-brand"
```
Comment (laurakwhit) [inline src/lib/components/schedule/schedule-form/spec-type-cron.svelte:93]: Can we use a Holocene component here? If not, we should add the type so it doesn't default to submit. 
```suggestion
    type="button"
    class="mr-auto underline hover:text-brand"
```

File: src/lib/components/schedule/schedule-form/schedule-details-card.svelte [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+          value="after"
+          label={translate('common.after')}
+        />
+        {#if $form.endKind === 'after'}
+          <div>
+            <Input
+              id="endAfterOccurrences"
+              type="number"
+              inputmode="numeric"
+              pattern="0-9+"
```
Comment (laurakwhit) [inline src/lib/components/schedule/schedule-form/schedule-details-card.svelte]: Looks like maybe this is an invalid regex? 
```suggestion
              pattern="[0-9]+"
```

File: src/lib/components/schedule/schedule-form/schedule-details-card.svelte [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+      <Input
+        id="jitter"
+        label={translate('schedules.jitter')}
+        type="number"
+        step={1}
+        min={0}
+        suffix={translate('common.seconds-abbreviated')}
+        bind:value={
+          () => $form.jitter.toString(), (v) => ($form.jitter = Number(v))
```
Comment (laurakwhit) [inline src/lib/components/schedule/schedule-form/schedule-details-card.svelte]:
```suggestion
          () => $form.jitter?.toString() ?? '', (v) => ($form.jitter = Number(v))
```

File: src/lib/components/schedule/constants.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  },
+  AllowAll: {
+    label: translate('schedules.overlap-allow-all-label'),
+    description: translate('schedules.overlap-allow-all-description'),
+  },
+});
+
+const MONTHS_UNIT: Unit<'month(s)'> = {
+  label: 'month(s)',
+  convert: (n: number) => n * 31 * 24 * 60 * 60,
```
Comment (laurakwhit) [inline src/lib/components/schedule/constants.ts:45]: Any potential issues with hard coding this to 31 days 🤔 Maybe we can hint at this in the UI somewhere...

File: src/lib/holocene/duration-input/duration-input.svelte [hunk 5]
```diff
@@ -79,16 +89,32 @@
...
+    if (!durationValue) return '';
+    const seconds = Number(parseDuration(durationValue));
+    if (!Number.isFinite(seconds)) return '';
+    const unitDef = units.find((u) => u.label === unitLabel);
+    const factor = unitDef ? unitDef.convert(1) : 1;
+    return factor ? String(seconds / factor) : String(seconds);
+  };
+
+  // svelte-ignore state_referenced_locally
+  let rawValue = $state(toUnitValue(value, initialUnit));
```
Comment (laurakwhit) [inline src/lib/holocene/duration-input/duration-input.svelte:116]: Does the `rawValue` get updated if the `value` changes outside of the component?

File: src/lib/components/schedule/utilities/get-form-spec.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+export function getFormSpecFromSpec(
+  spec: ScheduleSpec | null | undefined,
+): FormSpecSchema[] {
+  const specs: FormSpecSchema[] = [];
+
+  for (const calendar of spec?.structuredCalendar ?? []) {
+    specs.push({
+      kind: 'frozen',
```
Comment (laurakwhit) [inline src/lib/components/schedule/utilities/get-form-spec.ts:22]: Just want to confirm, once you create a calendar spec you can never edit it (even on create). Is that expected?
Image

File: src/lib/components/schedule/utilities/get-form-schedule-defaults.ts [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    // new schedules and off for existing ones, whose payloads aren't decoded
+    // back into the form.
+    editInput: !describeFullSchedule,
+    encoding: 'json/plain',
+    messageType: '',
+
+    specs: getSpecs(describeFullSchedule),
+    timezoneName: spec?.timezoneName || 'UTC',
+    startTime: isoStringToCalendarDateStr(
+      String(spec?.startTime || new Date().toISOString()),
```
Comment (laurakwhit) [inline src/lib/components/schedule/utilities/get-form-schedule-defaults.ts:144]: Is there potential for a bug here by setting the `startTime` when the user opens the form vs. submits it (e.g. if they submit a minute after midnight)? Seems like a very small likelihood, but thought it might be worth mention 😅

File: src/lib/components/schedule/schedule-form/schedule-policies-drawer.svelte [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+
+  // svelte-ignore state_referenced_locally
+  const policiesForm = superForm(pickPolicies(get(form)), {
+    SPA: true,
+    dataType: 'json',
+    validators: zodClient(formSchedulePoliciesSchema),
+    resetForm: false,
+    onUpdate: ({ form: validated }) => {
+      if (!validated.valid) return;
+      form.update(($form) => ({ ...$form, ...validated.data }));
```
Comment (laurakwhit) [inline src/lib/components/schedule/schedule-form/schedule-policies-drawer.svelte:71]: When I change any of these values and hit "Update Policies" it looks like the values update for a second on the "Schedule Policies" card and then the whole form reloads and the values are lost.

File: src/lib/components/schedule/utilities/get-form-spec.ts (hunk 6 above)
Comment (tegan-temporal) [reply src/lib/components/schedule/utilities/get-form-spec.ts:22]: yeah, you have to delete it and create a different one

File: src/lib/components/schedule/utilities/get-form-schedule-defaults.ts (hunk 7 above)
Comment (tegan-temporal) [reply src/lib/components/schedule/utilities/get-form-schedule-defaults.ts:144]: we set the startTime when the user opens the form so they see it in the form. I think it would be worse if we used a date different from what they see in the form. Also, `isoStringToCalendarDateStr` strips the time portion off, we're just using the date portion. Lastly, I don't think it necessarily matters at all if startDate is in the past. We don't force them to have startDate in the present or future. It just won't effect anything before the schedule is created.

File: src/lib/components/schedule/constants.ts (hunk 4 above)
Comment (tegan-temporal) [reply src/lib/components/schedule/constants.ts:45]: FWIW it is in the tooltip. I don't think there is a better solution besides just not supporting months (which I would be fine with, but design seemed to think it's still useful like this.
Image

Comment (laurakwhit) [reply src/lib/components/schedule/constants.ts:45]: Ah missed that tooltip. Ignore my comment then 😅

File: src/lib/holocene/duration-input/duration-input.svelte (hunk 5 above)
Comment (tegan-temporal) [reply src/lib/holocene/duration-input/duration-input.svelte:116]: no, it doesn't. The reason is because it also previously didn't and changing that behavior seemed like opening a can of worms since duration-input is pretty widely used and it would potentially change values that other parts of the app were receiving. Schedules doesn't need value to be controlled

File: src/lib/components/schedule/utilities/get-form-schedule-defaults.ts (hunk 7 above)
Comment (laurakwhit) [reply src/lib/components/schedule/utilities/get-form-schedule-defaults.ts:144]: That's fair. Is there a likely a scenario on edit where there **wouldn't** be a `startTime` defined? I see [your comments](https://github.com/temporalio/ui/pull/3603/changes#diff-83238a2cb660cf3e129e2a5d65fd63f1397ae7ddb5f2d87f52d6e85df9a8f918R172-R181) in `get-request-body.ts` and was mostly thinking that the conditional there would fail if the submit happened a "day" after the form was opened.

Comment (tegan-temporal) [reply src/lib/components/schedule/utilities/get-form-schedule-defaults.ts:144]: the scenario I was thinking for startTime being either missing or more specific than just a date is if you were using the ui to edit something that was created via the CLI

Comment (tegan-temporal) [reply src/lib/components/schedule/utilities/get-form-schedule-defaults.ts:144]: and I think I accounted for missing time just because of the way the protos have all fields as optional. You can't assume basically any value is really provided which is rather unfortunate.

Comment (laurakwhit) [reply src/lib/components/schedule/utilities/get-form-schedule-defaults.ts:144]: Yeah 😞 We can totally just call this a potential extreme edge case and leave as-is.

File: src/lib/components/schedule/utilities/get-request-body.test.ts [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+  it('emits the cron string for cron specs', async () => {
+    const body = await getRequestBody(buildForm());
+
+    expect(body.schedule.spec.cronString).toEqual(['0 12 * * *']);
+  });
+
+  it('drops invalid cron specs', async () => {
+    const body = await getRequestBody(
+      buildForm({ specs: [{ kind: 'cron', cronString: '0 12 * * *' }] }),
```
Comment (laurakwhit) [inline src/lib/components/schedule/utilities/get-request-body.test.ts]: It looks like this is a valid cron string. Should we be testing a invalid one here?

File: src/lib/components/schedule/schedule-form/schedule-details-card.svelte [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+              inputmode="numeric"
+              pattern="[0-9]+"
+              label={translate('schedules.occurrences-label')}
+              labelHidden
+              bind:value={
+                () => $form.endAfterOccurrences?.toString(),
+                (v) => ($form.endAfterOccurrences = Number(v))
+              }
+              placeholder={translate('schedules.occurrences-placeholder')}
+              min={1}
```
Comment (laurakwhit) [inline src/lib/components/schedule/schedule-form/schedule-details-card.svelte]: Looks like the there isn't an error for "0 occurrences", which means an unlimited schedule. Do we want to guard against that? 

Image

File: src/lib/components/schedule/schedule-action-modals/trigger-schedule-modal.svelte [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    namespace: string;
+    scheduleOverlapPolicy: OverlapPolicy;
+  }
+
+  const overlapPolicyContent = getOverlapPolicyContent();
+  const identity = getIdentity();
+
+  let { scheduleId, namespace, scheduleOverlapPolicy }: Props = $props();
+
+  let selectedOverlapPolicy = writable<OverlapPolicy>('AllowAll');
```
Comment (laurakwhit) [inline src/lib/components/schedule/schedule-action-modals/trigger-schedule-modal.svelte]: Should this reflect the `scheduleOverlapPolicy` instead of always defaulting to `AllowAll`?

**Review by laurakwhit — COMMENTED** [review]
A few more small comments, but overall lgtm 🎉

File: src/lib/components/schedule/schedule-action-modals/trigger-schedule-modal.svelte (hunk 11 above)
Comment (tegan-temporal) [reply src/lib/components/schedule/schedule-action-modals/trigger-schedule-modal.svelte]: since this is a manual trigger, IMO it is expected that if I trigger it, it runs, which isn't really guaranteed a lot of the other policies. Anyway, I will move it back to the schedule's specified policy as the default as that was the original behavior.

**Review by laurakwhit — APPROVED** [review]
